presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/InMemoryHashAggregationBuilder.java

```diff
@@ -284,6 +277,31 @@ protected Page computeNext()
         return types;
     }
 
+    /**
+     * Update memory usage with extra memory needed.
+     * The returned value indicates whether the reservation is within the limit
+     */
+    private boolean updateMemoryWithReservation(long reservedMemoryBytes)
+    {
+        long memorySize = getSizeInMemory() + reservedMemoryBytes;
+        if (partial) {
+            systemMemoryContext.setBytes(memorySize);
+            full = (memorySize > maxPartialMemory);
+        }
+        else {
+            if (!operatorContext.trySetMemoryReservation(memorySize)) {
+                operatorContext.setMemoryReservation(memorySize);
+                return false;
+            }
+            // trySetMemoryReservation will succeed if we have set memory reservation before
+            // cannot proceed to allocate more memory if we are still waiting for memory
+            if (!operatorContext.isWaitingForMemory().isDone()) {
+                return false;
```

Comment:
I wish it becomes sufficient that `trySetMemoryReservation` returns false and we stop working until more memory available. Then you don't need to inspect `isWaitingForMemory` which is kind of weird -- it should be suffficient for operator to declare memory usage and requirements and memory management should do the rest.